File: completed.rb
# script to update list of completed problems

completed_problems = {}

regex = /.+\/[A-Za-z_]*([0-9]+)\..+/

file_whitelist = %w(.git .gitignore files libraries COMPLETED FINISHED README TODO completed.rb)

language_directories = Dir['*']
language_directories -= file_whitelist

language_directories.each do |lang|
	files = Dir[lang + "/*"]
	files.each do |file|
		if file =~ regex then
			problem_num = file.match(regex).captures[0]
			if problem_num.length < 3 then
				problem_num = '0' * (3 - problem_num.length) + problem_num
			end
			completed_problems[problem_num] ||= [] # if doesn't exist yet, create an array, else do nothing
			completed_problems[problem_num].uniq!
			completed_problems[problem_num] << lang
		end
	end
end

File.open('COMPLETED', 'w') do |file|
	completed_problems.keys.sort.each do |key|
		file.write(key + ': ' + completed_problems[key].sort.join(', '))
		file.write("\n")
	end
	file.write("\nTotal: " + completed_problems.size.to_s + " challenges in " + language_directories.size.to_s + " languages.\n")
end

File: completed.py
import os
import re


IGNORELIST = {
    '.git',
    '.gitignore',
    'files',
    'libraries',
    'README.md',
    'COMPLETED',
    'TODO',
    'completed.py',
}

PROBLEM_REGEX = re.compile(r'.*?(\d+).*?')


def collect_problems(lang_dir):
    problems = set()
    with os.scandir(lang_dir) as it:
        for entry in it:
            name = str(entry.name)
            m = PROBLEM_REGEX.search(name)
            try:
                problems.add(int(m.group(1)))
            except (AttributeError, ValueError):
                continue
    return problems


def main():
    completed = {}
    languages = set()
    with os.scandir('.') as it:
        for entry in it:
            name = str(entry.name)
            if name in IGNORELIST:
                continue
            languages.add(name)
            for p in collect_problems(entry.path):
                if p not in completed:
                    completed[p] = []
                completed[p].append(name)

    with open('COMPLETED', 'w') as f:
        for p in sorted(completed.keys()):
            langs = ', '.join(sorted(completed[p]))
            f.write(f'{p:03d}: {langs}\n')

        f.write(f'\nTotal: {len(completed)} problems in {len(languages)} languages.')


if __name__ == '__main__':
    main()